lib/ansible/playbook/role/__init__.py

```diff
@@ -460,17 +472,30 @@ def get_dep_chain(self):
             dep_chain.append(parent)
         return dep_chain
 
+    @property
+    def cached_deps(self):
+        return global_role_dependency_cache.get(self.full_qualified_name, None)
+
+    def add_deps_to_cache(self, deps):
+        global_role_dependency_cache[self.full_qualified_name] = deps
+
     def get_default_vars(self, dep_chain=None):
         dep_chain = [] if dep_chain is None else dep_chain
 
-        default_vars = dict()
-        for dep in self.get_all_dependencies():
-            default_vars = combine_vars(default_vars, dep.get_default_vars())
-        if dep_chain:
-            for parent in dep_chain:
-                default_vars = combine_vars(default_vars, parent._default_vars)
-        default_vars = combine_vars(default_vars, self._default_vars)
-        return default_vars
+        deps = self.cached_deps
+        if deps is None:
+            deps = self.get_all_dependencies()
+            self.add_deps_to_cache(deps)
+
+        if self._default_vars_full is None:
+            self._default_vars_full = dict()
+            for dep in deps:
+                self._default_vars_full = combine_vars(self._default_vars_full, dep.get_default_vars())
+            if dep_chain:
+                for parent in dep_chain:
+                    self._default_vars_full = combine_vars(self._default_vars_full, parent._default_vars)
+            self._default_vars_full = combine_vars(self._default_vars_full, self._default_vars)
```

Comment:
Looks like this chunk could also be a `@cached_property` if I'm reading this right. And with that, we wouldn't even need the `self._default_vars_full = None` definition in `__init__()` to begin with.